1552_ek_certificate_chain.md

```diff
@@ -0,0 +1,147 @@
+
+## Design Details
+First some words from TCG EK documentations:
+
+```txt
+TPM NV MAY contain all certificates of the EK Certificate Chain except the Root CA certificate. The
+EK Certificate Chain MUST be stored as X.509 DER encoded certificates. If the chain consists of
+more than one certificate, or if multiple chains exist, they MUST be stored in NV as a concatenated
+sequence. The TPM manufacturer MAY provision certificate chains using the following list of indices:
+
+0x01c00100 - 0x01c001ff
+
+Revision 2 January 26, 2022 Published Page 15 of 60
+The NV indices MUST be populated starting with index 0x01c00100 through index 0x01c001ff. There
+is no requirement to store the certificates in any particular order whatsoever. If a concatenated
+certificate chain does not fit in a single index, the chain MUST overflow to the next numerically larger
+index in the list of NV Indices. If the storage space in a single index is insufficient to store the entire
+certificate, the certificate MAY overflow into the next numerically larger index in the list of NV Indices.
+```
+
+The feature requires changes in the rust-keylime agent and in the keylime servers registrar and tenant. The certificate chain must and can be only provided by the keylime-agent.
+
+The keylime-agent must do the following:
+* get all available NV-Index-Handles in the defined range (`0X01C00100 - 0X01C001FF`)
+* read the chain from the NV-Index-Handles
+  * Theoretically the certificates can be in any order. This would require an option to make a sorting possible in the agent. As it does not make really sense to store the chain in an unsorted order i would postpone any implementation till this case really exists.
+  * Only a subset of the NV-Index-Handles may be required. This would require an option to filter/set the required handles. I would also wait for implementation till a real case exists for that.
+* as the certificate chain is stored in DER format, first the certificates must be separated
+* transform der DER certificates into PEM format.
+* send the chain to the keylime server
+
+The certificate chain should be transformed and sorted by the keylime-agent to keep the work
+away from the server and provide a generic server implementation that does not need to behave
+a certain way (sort/filter the chain) depending on the client.
+
+The keylime registrar only needs to store the certificate chain in the database.
+
+The keylime tenant must detect the presence of an EK Certificate Chain and verify the
+EK Cert against the chain, and finally the top level certificate against the `tpm_cert_store`.
+
+The flow can be kept mostly as it is in `check_ek`. In case of a present ek_ca_chain, the ek must be verified against the provided chain and in case of a success the ek will be replaced by the top level certificate from the chain. Afterwards the flow can be kept as it was.
```

Comment:
```suggestion
The flow can be kept mostly as it is in `check_ek`. In case of multiple certs in `ekcert` , the ek must be verified against the provided chain and in case of a success the ek will be replaced by the top level certificate from the chain. Afterwards the flow can be kept as it was.
```